src/rsa/blinding.rs

```diff
@@ -0,0 +1,644 @@
+        // [NSA Suite B Implementer's Guide to NIST SP 800-56A] Appendix B.2,
+        // "Key Pair Generation by Testing Candidates".
+        //
+        // [NSA Suite B Implementer's Guide to ECDSA]: doc/ecdsa.pdf.
+        // [NSA Suite B Implementer's Guide to NIST SP 800-56A]: doc/ecdh.pdf.
+
+        assert_eq!(self.max_exclusive.len(), dest.len());
+
+        // XXX: The value 100 was chosen to match OpenSSL due to uncertainty of
+        // what specific value would be better, but it seems bad to try 100
+        // times.
+        for _ in 0..100 {
+            {
+                let mut dest_as_bytes = limbs_as_bytes_mut(dest);
+                try!(rng.fill(&mut dest_as_bytes));
+            }
+
+            {
+                // Mask off unwanted bits
+                let mask = self.sampling_params.most_sig_limb_mask;
+                dest[self.max_exclusive.len() - 1] &= mask;
+            }
+
+            if self.are_limbs_within(&dest) {
+                return Ok(());
+            }
+
+            if self.sampling_params.reduce_when_over_bound {
+                // `dest` is not in (0, max] but maybe we can fix that.
+                // (See above for explanation of why this is safe.)
+
+                limbs_reduce_once_constant_time(dest, self.max_exclusive);
+                if self.are_limbs_within(&dest) {
+                    // `dest` was in [max, 2*max)
+                    return Ok(());
+                }
+
+                limbs_reduce_once_constant_time(dest, self.max_exclusive);
+                if self.are_limbs_within(&dest) {
+                    // `dest` was in [2*max, 3*max)
+                    return Ok(());
+                }
+
+                // `dest` was in [3*max, 2**(n+1)) or zero
+            }
+        }
+
+        Err(error::Unspecified)
+    }
+}
+
+#[allow(unsafe_code)]
+#[allow(non_snake_case)]
+#[doc(hidden)]
+#[no_mangle]
+pub unsafe extern fn GFp_rand_mod(dest: *mut Limb, max_exclusive: *const Limb,
+                                  num_limbs: c::size_t, rng: *mut rand::RAND)
+                                  -> c::int {
+    const ERR: c::int = 0;
+    const SUCCESS: c::int = 1;
+
+    let max_exclusive = core::slice::from_raw_parts(max_exclusive, num_limbs);
+    let mut dest = core::slice::from_raw_parts_mut(dest, num_limbs);
+
+    let result = set_to_rand_mod(&mut dest, &max_exclusive, (*rng).rng);
+    if result.is_err() {
+        return ERR;
+    }
+
+    SUCCESS
+}
+
+/// Chooses a positive integer less than `max_exclusive` uniformly at random
+/// and stores it into `dest`.
+fn set_to_rand_mod(dest: &mut [Limb], max_exclusive: &[Limb], rng: &rand::SecureRandom)
+            -> Result<(), error::Unspecified> {
+
+    debug_assert_eq!(dest.len(), max_exclusive.len());
+
+    let sampling_params = select_sampling_params(max_exclusive);
+
+    if sampling_params.extend_limbs_by_one {
+        with_extended_buffers(dest, max_exclusive, |ext_dest, ext_max| {
```

Comment:
I suggest that you always just copy the stuff into the extended buffers and then just inline `with_extended_buffers` into this function. The most common case is a 2048-bit modulus that starts with `0b100.....`, right? If that case is too slow then we'd need to fix things. Conversely, if that case isn't to slow then there's no point in optimizing anything about the other case. In general, I feel like this is pretty over-engineered. I suggest trying to write it the simplest possible way that works for the 2048-bit case where the public modulus starts with `0b100...`, and then do the simplest possible (if suboptimal) thing for all other cases.